esp-hal/src/rsa/mod.rs

```diff
@@ -373,49 +372,61 @@ pub(crate) mod asynch {
     use core::task::Poll;
 
     use embassy_sync::waitqueue::AtomicWaker;
+    use portable_atomic::{AtomicBool, Ordering};
     use procmacros::handler;
 
-    use crate::{Async, rsa::{
-        Multi,Rsa,
-        RsaMode,
-        RsaModularExponentiation,
-        RsaModularMultiplication,
-        RsaMultiplication,
-    }};
+    use crate::{
+        rsa::{
+            Multi,
+            Rsa,
+            RsaMode,
+            RsaModularExponentiation,
+            RsaModularMultiplication,
+            RsaMultiplication,
+        },
+        Async,
+    };
 
     static WAKER: AtomicWaker = AtomicWaker::new();
 
+    static SIGNALED: AtomicBool = AtomicBool::new(false);
+
     /// `Future` that waits for the RSA operation to complete.
     #[must_use = "futures do nothing unless you `.await` or poll them"]
-    pub(crate) struct RsaFuture<'d> {
-        instance: &'d crate::peripherals::RSA,
+    struct RsaFuture<'a, 'd> {
+        #[cfg_attr(esp32, allow(dead_code))]
+        instance: &'a Rsa<'d, Async>,
     }
 
-    impl<'d> RsaFuture<'d> {
-        /// Asynchronously initializes the RSA peripheral.
-        fn new(instance: &'d crate::peripherals::RSA) -> Self {
+    impl<'a, 'd> RsaFuture<'a, 'd> {
+        fn new(instance: &'a Rsa<'d, Async>) -> Self {
+            SIGNALED.store(false, Ordering::Relaxed);
+
             cfg_if::cfg_if! {
                 if #[cfg(esp32)] {
-                    instance.rsa.interrupt().modify(|_, w| w.interrupt().set_bit());
                 } else if #[cfg(any(esp32s2, esp32s3))] {
-                    instance.rsa.interrupt_ena().modify(|_, w| w.interrupt_ena().set_bit());
+                    instance.rsa.interrupt_ena().write(|w| w.interrupt_ena().set_bit());
                 } else {
-                    instance.rsa.int_ena().modify(|_, w| w.int_ena().set_bit());
+                    instance.rsa.int_ena().write(|w| w.int_ena().set_bit());
                 }
             }
 
             Self { instance }
         }
 
-        fn event_bit_is_clear(&self) -> bool {
-            // Looks weird, but we signal completion by disabling the interrupt.
+        fn is_done(&self) -> bool {
+            SIGNALED.load(Ordering::Relaxed)
```

Comment:
This should definitely be `Acquire`

Reply:
I don't claim to understand memory orderings, but my understanding is that `Relaxed` should be safe here, it's also used in Mara's [book](https://marabos.nl/atomics/atomics.html#example-stop-flag) in an example of similar semantics.

I'd be happy to learn *why* this definitely has to be something else, if you would please take the time to explain or point me in the right direction.

Reply:
The best reference I have is https://doc.rust-lang.org/std/sync/atomic/enum.Ordering.html#variant.Acquire and possibly some Jon Gjengset videos on YouTube.

The reason to use `Acquire` here is users of `is_done` are relying on the invariant that if it returns true, all the stuff that they expect to be done has been done. Without `Acquire`, it's possible that it returns true before stuff is done.

Of course this needs to be paired with a `Release` in the interrupt handler to make sure everything is done before the boolean is marked true.

Lemme know if you want more words.

Reply:
> Without Acquire, it's possible that it returns true before stuff is done.

As the counterpart is in an interrupt handler, I think we can be confident everything will be done by the time this line executes and returns true 🤔

Reply:
Yes that's one half of the equation. (We're assuming single core interrupts of course)

```Rust
while !fut.is_done() {}

// do other stuff.
// Without Acquire, the compiler (and CPU in theory, I don't think ESP32s do reordering) could move this to before the while loop.
```